src/storage/src/controller.rs

```diff
@@ -984,6 +955,221 @@ impl<T: Timestamp> CollectionState<T> {
+                                        id
+                                    );
+                                }
+                            }
+                            PersistWorkerCmd::Downgrade(since_frontiers) => {
+                                for (id, frontier) in since_frontiers {
+                                    downgrades.insert(id, frontier);
+                                }
+                            }
+                            PersistWorkerCmd::Snapshot(id, as_of, oneshot) => {
+                                async fn snapshot<T2: Timestamp + Lattice + Codec64>(
+                                    read_handle: &mut ReadHandle<SourceData, (), T2, Diff>,
+                                    id: GlobalId,
+                                    as_of: Antichain<T2>,
+                                ) -> Result<Vec<(Row, Diff)>, StorageError>
+                                {
+                                    let mut snapshot = read_handle
+                                        .snapshot(as_of)
+                                        .await
+                                        .map_err(|_| StorageError::ReadBeforeSince(id))?;
+
+                                    let mut contents = Vec::new();
+
+                                    while let Some(updates) = snapshot.next().await {
+                                        for ((source_data, _pid), _ts, diff) in updates {
+                                            let row =
+                                                source_data.expect("cannot read snapshot").0?;
+                                            contents.push((row, diff));
+                                        }
+                                    }
+
+                                    Ok(contents)
+                                }
+
+                                if let Some(read_handle) = read_handles.get_mut(&id) {
+                                    let result = snapshot(read_handle, id, as_of).await;
+                                    oneshot.send(result).expect("Oneshot should not fail");
+                                } else {
+                                    oneshot
+                                        .send(Err(StorageError::IdentifierMissing(id)))
+                                        .expect("Oneshot should not fail");
+                                }
+                            }
+                            PersistWorkerCmd::Shutdown => {
+                                shutdown = true;
+                            }
+                        }
+                    }
+
+                    if !downgrades.is_empty() {
+                        let futs = FuturesUnordered::new();
+
+                        for (id, read) in read_handles.iter_mut() {
+                            if let Some(since) = downgrades.remove(id) {
+                                let fut = async move {
+                                    read.downgrade_since(since).await;
```

Comment:
should this delete the read handle after the downgrade if the since is the empty antichain (so we don't keep handles around for deleted shards)?

Reply:
ah idempotency is a good point, nevermind then. as far as having fewer of them around to help out persist, if some ReadHandle has processed a downgrade_since of the empty antichain, then... there's not really anything it can do after that. persist should probably just internally make that inexpensive